protocol/deployment-chains.md

```diff
@@ -0,0 +1,261 @@
+
+Not all that different from status-quo.
+With copy-paste we can add 1 more chain, and suffix it with `-interop`, to produce a 2-chain test environment.
+For early testing this was drafted in [PR 10777](https://github.com/ethereum-optimism/optimism/pull/10777).
+
+This does not yet unify the common superchain part of the deployment, or any of the artifacts,
+nor supports test-specific changes to parameters.
+
+## Presets and configs
+
+In Eth2 there exist two classes of configuration:
+- "preset" (named, enshrined, compile-time)
+- "config" (customizable, run-time)
+
+This helped reduce complexity by encapsulating and limiting compile-time configuration,
+but also limited severely what could be changed.
+For a state-format it may be good to contain options,
+but for the amount of runtime configuration the OP-Stack has, it seems impractical.
+
+And each variable that remains in the runtime-config still cannot be customized in `op-e2e` tests,
+due to limits in the current `l1-allocs` and `l2-allocs` pre-generated files approach.
+
+## Templating
+
+By separating contract-code and contract-storage, the contract deployments can be reconfigured without recompiling.
+However, this pushes complexity to the user of the deployments data:
+it re-introduces smart-contract deployment concepts into the Go chain-ops.
+E.g. storage-layouts, storage packing, the occasional `immutable` variable, all make customization difficult.
+Not to mention cases where we need entirely different contracts, different hardforks, or more L2 deployments.
+
+The insight here is that we should strongly contain smart-contract configuration to the solidity-scripting side,
+and minimize post-processing in Go.
+
+## Mega test deployment
+
+Instead of generating many separate files per configuration,
+all deployments could be applied to a single `allocs` with all contracts, and many different parameter choices.
+It becomes bulky, but then allows the consumer to selectively choose what
+part of the deployment matters, and what is dead world state.
+
+This falls apart because the consumer has to track all the deployments,
+and it still leaves the issues of enshrined test-configurations.
+
+# Proposed Solution
+
+### Requirements
+
+The solution requirements boil down to:
+- Enable full configurability.
+- Make configuration composable, to have multiple L2s, and separate "superchain" from "chain".
+- Generate the deployments and L2 state with solidity scripts.
+
+This implies:
+- We need to invoke chain deployment during test runtime, since it cannot be captured ahead of time.
+- We need the deployment and configuration to be split, such that we can compose the steps.
+
+#### Addressing gaps
+
+##### How do we make this practical (fast) in tests?
+
+We can run it once, and cache the result, if it is deterministic, and if time-specifics can be adjusted later.
+
+##### How do we make this composable?
+
+We group the inputs, per deployment step, to not do everything at once.
```

Comment:
This is great. One thing I really don't like about the current deploy config is that all types of inputs are treated equally—some inputs are set once at genesis and never change, others are config params that can be changed anytime by a privileged role, etc. A clean grouping of inputs will be very helpful